### PR TITLE
Lift rpc_timeout to RpcAgent, for other RpcAgents to reuse.

### PR DESCRIPTION
**Pull request by xush6528**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29341 Lift rpc_timeout to RpcAgent, for other RpcAgents to reuse.**

https://github.com/pytorch/pytorch/pull/29336

So that other RpcAgent could use this timeout setting as well.

Differential Revision: [D5681951](https://our.internmc.facebook.com/intern/diff/D5681951/)

### CODE REVIEW

**Review by pietern — COMMENTED** [review]
@pritamdamania87 commented in the original PR https://github.com/pytorch/pytorch/pull/29336#issuecomment-550585781 about why there should be 3 timeouts and I agree. From the client's perspective there is no difference between the phases you list. Allowing different timeouts to be used means it's practically impossible to reason about timeouts (or spot them in log output) anymore.

Looking at the contents of the PR, I see two things:
1. Lifting `rpcTimeout` from ProcessGroupAgent into RPCAgent and renaming it
2. Unrelated formatting changes

We can all agree that lifting the parameter to RPCAgent is a good idea. Renaming is a separate issue and assumes we're going to have a bunch of timeouts, so we should discuss this more before going through with it. The formatting changes should be a separate formatting-only PR so it keeps the history usable (or you'll end up investigating why those lines were touched as part of a timeout change).

I propose the following:
1. Remove the rename from this PR and make it only move the timeout field to the base class.
2. Submit an issue for the rename so we can discuss what you're proposing.
3. Submit a separate PR for the formatting changes.

**Comment by xush6528** [comment]
@pietern For 2), opened #29402 to explain the whole context and idea.

For 1), 3), reverting in this PR.

File: torch/csrc/distributed/rpc/rpc_agent.h
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rpc_agent.h]: Why this needs to be virtual. In which situation a derived class would like to override this value?

Can we make it inline?

File: torch/distributed/rpc/__init__.py
Comment (mrshenli) [inline torch/distributed/rpc/__init__.py]: Before we made the decision on whether we should have finer grain timeout (server, client, etc.), shall we avoid mentioning server here?

File: torch/csrc/distributed/rpc/init.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp]: Question, if the RpcAgent will handle timeout internally, why do we need to expose this API in Python and C++? Is this only for test and debugging?

File: torch/csrc/distributed/rpc/rpc_agent.h
Comment (xush6528) [reply torch/csrc/distributed/rpc/rpc_agent.h]: Removing it.

Marking it as inline.

Turned out pybind11 does not require it to be declared in sub-classes. pybind11 is able to look up this member function in the base class.

File: torch/distributed/rpc/__init__.py
Comment (xush6528) [reply torch/distributed/rpc/__init__.py]: Makes sense. Removing it for now.

File: torch/csrc/distributed/rpc/init.cpp
Comment (xush6528) [reply torch/csrc/distributed/rpc/init.cpp]: I think what @rohan-varma mentioned was since we allow users to set timeout, we need to have a getter to allow users to inspect this value as well.

Comment (mrshenli) [reply torch/csrc/distributed/rpc/init.cpp]: I see, makes sense

**Review by mrshenli — APPROVED** [review]
LGTM! Approve to unblock. Please make sure tests pass before landing.

**Comment by pietern** [comment]
@xush6528 Thanks!

File: torch/distributed/rpc/api.py
Comment (pritamdamania87) [inline torch/distributed/rpc/api.py]: nit: I don't think we should have backticks (`) in exception messages. This would be displayed on the user's terminal and there is no markdown there.

File: test/rpc_test.py
Comment (pritamdamania87) [inline test/rpc_test.py]: See the discussion in https://github.com/pytorch/pytorch/pull/28926, we should directly add docstrings to the pybind method.

Comment (rohan-varma) [reply test/rpc_test.py]: @pritamdamania87 currently, `ProcessGroupAgent` is not exposed in the docs since I assumed that this is a private class and there's no reason for the user to directly access it (we provide higher level APIs for everything that needs to be done). If we wanted to move this doc to pybind, we'd have to expose documentation for `ProcessGroupAgent`, since `get_rpc_timeout` is a method on that class.

Comment (xush6528) [reply test/rpc_test.py]: Created issue #29628 for tracking thihs.

File: torch/distributed/rpc/api.py
Comment (xush6528) [reply torch/distributed/rpc/api.py]: @pritamdamania87 

I was not expecting a markdown format. But want to make clear that it's a python variable. Any other suggestion to make it distinct? or you think plain text is fine?

File: test/rpc_test.py
Comment (pritamdamania87) [reply test/rpc_test.py]: Can we move this method from `RpcAgent` to `torch.distributed.rpc` and the implementation uses the default RpcAgent?

**Pull request closed without merging**
